lgc/patch/PatchInitializeWorkgroupMemory.cpp

```diff
@@ -107,65 +118,146 @@ bool PatchInitializeWorkgroupMemory::runOnModule(Module &module) {
-void PatchInitializeWorkgroupMemory::initializeWithZero(Value *pointer, Type *valueTy,
-                                                        SmallVectorImpl<Value *> &indices, BuilderBase &builder) {
-  if (valueTy->isSingleValueType()) {
-    Value *zero = Constant::getNullValue(valueTy);
-    unsigned alignment = valueTy->getPrimitiveSizeInBits() / 8;
-    if (!isPowerOf2_32(alignment))
-      alignment = NextPowerOf2(alignment);
-    if (!indices.empty())
-      pointer = builder.CreateGEP(pointer, indices);
-    builder.CreateAlignedStore(zero, pointer, Align(alignment));
-    return;
-  } else if (valueTy->isArrayTy()) {
-    const unsigned elemCount = valueTy->getArrayNumElements();
-    valueTy = valueTy->getContainedType(0);
-    for (unsigned idx = 0; idx < elemCount; ++idx) {
-      indices.push_back(builder.getInt32(idx));
-      initializeWithZero(pointer, valueTy, indices, builder);
-      indices.pop_back();
+  auto initBlock = BasicBlock::Create(*m_context, ".init", originBlock->getParent(), endInitBlock);
+  auto bodyBlock = BasicBlock::Create(*m_context, ".body", originBlock->getParent(), initBlock);
+  auto forHeaderBlock = BasicBlock::Create(*m_context, ".for.header", originBlock->getParent(), bodyBlock);
+
+  builder.SetInsertPoint(originBlock->getTerminator());
+  // Get thread info
+  auto &shaderMode = m_pipelineState->getShaderModes()->getComputeShaderMode();
+  const auto &entryArgIdxs = m_pipelineState->getShaderInterfaceData(m_shaderStage)->entryArgIdxs;
+  Value *localInvocationId = getFunctionArgument(m_entryPoint, entryArgIdxs.cs.localInvocationId);
+  const unsigned numThreads = shaderMode.workgroupSizeX * shaderMode.workgroupSizeY * shaderMode.workgroupSizeZ;
+
+  Value *threadId = builder.CreateExtractElement(localInvocationId, uint64_t(0));
+  if (shaderMode.workgroupSizeY > 1) {
+    Value *stride = builder.CreateMul(builder.getInt32(shaderMode.workgroupSizeX),
+                                      builder.CreateExtractElement(localInvocationId, 1));
+    threadId = builder.CreateAdd(threadId, stride);
+  }
+  if (shaderMode.workgroupSizeZ > 1) {
+    Value *stride = builder.CreateMul(builder.getInt32(shaderMode.workgroupSizeX * shaderMode.workgroupSizeY),
+                                      builder.CreateExtractElement(localInvocationId, 2));
+    threadId = builder.CreateAdd(threadId, stride);
+  }
+  originBlock->getTerminator()->replaceUsesOfWith(endInitBlock, forHeaderBlock);
+
+  // Each thread stores a zero to the continues LDS
+  // for (int loopIdx = 0; loopIdx < loopCount; ++loopIdx) {
+  //   if (threadId x loopCount + loopIdx < requiredThreads) {
+  //      unsigned ldsOffset = (threadId * loopCount) + loopIdx;
+  //      CreateStore(zero, ldsOffset);
+  //   }
+  //  }
+
+  PHINode *loopIdxPhi = nullptr;
+  const unsigned requiredThreads = lds->getType()->getPointerElementType()->getArrayNumElements();
+  Value *zero = Constant::getNullValue(builder.getInt32Ty());
```

Comment:
Could be replaced with builder.getInt32(0) ?

Reply:
Done.